vcd/resource_vcd_vapp_vm.go

```diff
@@ -583,421 +591,741 @@ func vmSchemaFunc(vmType typeOfVm) map[string]*schema.Schema {
+//   - OS Type
+//   - Hardware version
+//   - Storage profile configuration
+//   - VM compute policy configuration
+//   - Boot image
+//
+// 2. Perform VM creation operation based on type in separate switch/case
+//   - standaloneVmType
+//   - vAppVmType
+//
+// # This part includes defining initial structures for VM and also any explicitly required operations for that type of VM
+//
+// 3. Perform additional operations which are common for both types of VMs
+//
+// Note. VM Power ON (if it wasn't disabled in HCL configuration) occurs as last step after all configuration is done.
+func createVmEmpty(d *schema.ResourceData, meta interface{}, vmType typeOfVm) (*govcd.VM, error) {
+	util.Logger.Printf("[TRACE] Creating empty VM: %s", d.Get("name").(string))
+
 	vcdClient := meta.(*VCDClient)
+	org, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return nil, fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
 
-	// When there is more then one VM in a vApp Terraform will try to parallelise their creation.
-	// However, vApp throws errors when simultaneous requests are executed.
-	// To avoid them, below block is using mutex as a workaround,
-	// so that the one vApp VMs are created not in parallelisation.
+	var vapp *govcd.VApp
 
 	if vmType == vappVmType {
-		vcdClient.lockParentVapp(d)
-		defer vcdClient.unLockParentVapp(d)
+		vappName := d.Get("vapp_name").(string)
+		vapp, err = vdc.GetVAppByName(vappName, false)
+		if err != nil {
+			return nil, fmt.Errorf("[VM create] error finding vApp for empty VM%s: %s", vappName, err)
+		}
 	}
 
-	// Exit early only if "network_dhcp_wait_seconds" is changed because this field only supports
+	var ok bool
+	var osType interface{}
+
+	if osType, ok = d.GetOk("os_type"); !ok {
+		return nil, fmt.Errorf("`os_type` is required when creating empty VM")
+	}
+
+	var hardWareVersion interface{}
+	if hardWareVersion, ok = d.GetOk("hardware_version"); !ok {
+		return nil, fmt.Errorf("`hardware_version` is required when creating empty VM")
+	}
+
+	var computerName interface{}
+	if computerName, ok = d.GetOk("computer_name"); !ok {
+		return nil, fmt.Errorf("`computer_name` is required when creating empty VM")
+	}
+
+	var bootImage *types.Media
+	if bootImageName, ok := d.GetOk("boot_image"); ok {
+		var catalogName interface{}
+		if catalogName, ok = d.GetOk("catalog_name"); !ok {
+			return nil, fmt.Errorf("`catalog_name` is required when creating empty VM with boot_image")
+		}
+		catalog, err := org.GetCatalogByName(catalogName.(string), false)
+		if err != nil {
+			return nil, fmt.Errorf("error finding catalog %s: %s", catalogName, err)
+		}
+		result, err := catalog.GetMediaByName(bootImageName.(string), false)
+		if err != nil {
+			return nil, fmt.Errorf("[VM creation] error getting boot image %s : %s", bootImageName, err)
+		}
+
+		bootImage = &types.Media{HREF: result.Media.HREF, Name: result.Media.Name, ID: result.Media.ID}
+	} else {
+		bootImage = nil
```

Comment:
I think this `else` clause can be removed, can't it?

Reply:
Yes!